homeassistant/components/google/quality_scale.yaml

```diff
@@ -0,0 +1,76 @@
+rules:
+  # Bronze
+  config-flow:
+    status: done
+    comment: Integration uses an OAuth config flow.
+  brands: done
+  dependency-transparency: todo
+  common-modules: done
+  has-entity-name: done
+  action-setup:
+    status: todo
+    comment: |
+      Actions are current setup in `async_setup_entry` and need to be moved
+      to `async_setup`.
+  appropriate-polling: done
+  test-before-configure: done
+  entity-event-setup:
+    status: exempt
+    comment: Integration does not subscribe to events.
+  unique-config-entry: done
+  entity-unique-id: done
+  docs-installation-instructions: done
+  docs-removal-instructions: todo
+  test-before-setup:
+    status: todo
+    comment: |
+      The integration does not test the connection in `async_setup` but
+      instead does this in the calendar platform only, which can be improved.
+  docs-high-level-description: done
+  config-flow-test-coverage: done
```

Comment:
`test_full_flow_application_creds` does not assert unique id

Reply:
Neither is `test_exchange_error` (afaict its also a full flow test, so this is relevant)

Reply:
Are you saying this shouldn't be marked 100% test coverage?

Reply:
The unique id is tested with `test_duplicate_config_entries`
https://github.com/home-assistant/core/blob/355e80aa56cf087f7b5b545e4209b2cb718eea87/tests/components/google_translate/test_config_flow.py#L68

Reply:
No, we only test that it's aborting because the entry already exists, it doesn't tell us anything about what the unique id is

Reply:
The test is verifying the unique id is the same as the one the test used for the existing config entry. It's not again asserting the actual string contents directly because it doesn't matter for the purposes of the test exactly what the unique id, just that the behaviors of unique ids are enforced (that you can't set up the same account twice). If there are other behaviors that need to be asserted we can add those.


That guidance says "make sure to also test that the flow doesn't allow adding more than one unique configuration entry to reach complete coverage" and this is doing it but if you want to update it to be more prescriptive here about what exactly you want to see, then i'll follow it, but i think we're getting into diminishing returns or style.

Reply:
Okay Now that I reread my message I was a little too short in my response.

Yes this test is currently testing that we can't add the same entry twice. That is good.

But I consider the set unique_id as a part of the end result, just like how we assert data of the config entry. Maybe in this integration it would not be easily possible, but one could refactor to use `self._async_abort_entries_match()` and this test would succeed, but the unique id would not be set. A more realistic reason why we would want this is that there are multiple flows that can lead to the same result. By also asserting the unique_id at both flows, we avoid the risk of refactoring a flow where it forgets to call the `self.abort_if_unique_id_set()`. (But now that I write this I also think that that will be caught again by this test).

So I think the TLDR is: I consider it a good practice to assert, just to be explicit. It probably won't be as much of a danger for google_calendar (as you only have tokens, not a host or something).